src/radix_connect/app_link_parser.rs

```diff
@@ -0,0 +1,173 @@
+use crate::prelude::*;
+use url::form_urlencoded;
+use url::Url;
+
+const CONNECT_URL_PARAM_SESSION_ID: &str = "sessionId";
+const CONNECT_URL_PARAM_ORIGIN: &str = "origin";
+const CONNECT_URL_PARAM_INTERACTION_ID: &str = "interactionId";
+const CONNECT_URL: &str = "https://d1rxdfxrfmemlj.cloudfront.net";
+
+pub fn parse_mobile_connect_request(
+    url: String,
+) -> Result<MobileConnectRequest> {
+    let connect_url = Url::parse(CONNECT_URL).unwrap();
+    let parsed_url = Url::parse(url.as_str()).map_err(|_| {
+        CommonError::RadixConnectMobileInvalidRequestUrl {
+            bad_value: url.clone(),
+        }
+    })?;
+    if parsed_url.host_str() != connect_url.host_str()
+        || parsed_url.scheme() != connect_url.scheme()
+    {
+        return Err(CommonError::RadixConnectMobileInvalidRequestUrl {
+            bad_value: url.clone(),
+        });
+    }
+
+    let query_parameters = parsed_url
+        .query_pairs()
+        .into_owned()
+        .collect::<HashMap<String, String>>();
+    let session_id_string = query_parameters
+        .get(CONNECT_URL_PARAM_SESSION_ID)
+        .ok_or(CommonError::RadixConnectMobileInvalidRequestUrl {
+            bad_value: url.clone(),
+        })?;
+    let session_id = SessionID::from_str(session_id_string).map_err(|_| {
+        CommonError::RadixConnectMobileInvalidRequestUrl {
+            bad_value: url.clone(),
+        }
+    })?;
+
+    match query_parameters.get(CONNECT_URL_PARAM_ORIGIN) {
```

Comment:
This last match here, can be reworked to use `let else` which makes a more readable, and this solution also scales better, moving heavy lifting done here *ad-hoc* into reusable code:

something like: 

```rust
{ ...

let Some(origin) = query_parameters.get(CONNECT_URL_PARAM_ORIGIN) else {
    query_parameters
        .get(CONNECT_URL_PARAM_INTERACTION_ID)
        .ok_or(CommonError::RadixConnectMobileInvalidRequestUrl {
            bad_value: url.clone(),
        })
        .and_then(|interaction_id| {
            DappRequest::try_with_interaction_id_and_session_id(
                interaction_id,
                session_id
            )
        })
        .map(MobileConnectRequest::DappInteraction)
};

LinkRequest::try_with_origin_and_session_id(origin, session_id)
        .map(MobileConnectRequest::Link)

}
```

Inside `link_request.rs` file

```rust
impl LinkRequest {
    pub(crate) fn try_with_origin_and_session_id(
        origin: impl AsRef<str>,
        session_id: SessionD
    ) -> Result<Self> {
        let origin = Url::parse(origin) // 🔮 you MIGHT have to call `.as_ref()` 
            .map_err(|_| CommonError::RadixConnectMobileInvalidOrigin {
                bad_value: origin.to_owned(),
            })?;

        OK(Self::new(origin, session_id)) // 🔮 write a constructor `new`
    }
}
```

and inside `dapp_request.rs` file

```rust
impl DappRequest {
    pub(crate) fn try_with_interaction_id_and_session_id(
        interaction_id: impl AsRef<str>,
        session_id: SessionID
    ) -> Result<Self> {

        // 🔮 WalletInteractionId is probably a Uuid...
        // https://github.com/radixdlt/sargon/pull/98/files#r1578324966
        // so must TRY parse it with `FromStr` (returns Result)
        let interaction_id = WalletInteractionId.from_str(interaction_id)?;
        Ok(Self::new(interaction_id, session_id)) // 🔮 write a constructor `new`
    }
}
```